### PR TITLE
spirv: test runner support

### PR DESCRIPTION
**Pull request by Snektron**
This PR adds support code for running SPIR-V tests. This consists of a whole bunch of stuff:
- Constant generation: general overhaul, support for many types.
  - Constant generation now ignores SPIR-V and lowers constants as an untyped byte bag.
- Type lowering: many types also.
- Code generation: switches, some arithmetic, calls, and more.
- start.zig support to export main functions. This is mostly only useful for test support.
- linking improvements for globals and functions, the backend can now properly order globals and generate OpEntryPoint instructions,
- Unifies PtxKernel and AmdgpuKernel into Kernel, and adds support for Kernels for OpenCL. Functions exported with the Kernel calling convention are automatically added as entry point.
- Temporary test kernel generation to allow us to at least run SOME tests. We cannot implement the standard library mechanism for this currently as it requires function pointers, and even though it is possible to emulate those, I would like to be able to work on the behavior tests a little bit before that.

Im happy to report that the goal is achieved: SPIR-V tests can now be executed! This requires a special implementation of the test executor and test runner, which can be found on my repository here: https://github.com/Snektron/zig-spirv-test-runner. This repo also has a brief description of the design and usage. Preview from that repo:
```
$ zig test src/test_kernel.zig --test-runner src/test_runner.zig -fno-compiler-rt -target spirv64-opencl -mcpu generic+Int64+Int16+Int8 --test-cmd zig-out/bin/zig-spirv-executor --test-cmd --platform --test-cmd Intel --test-cmd-bin
Test [2/3] test test_kernel.test.fail... FAIL (Fail)
Test [3/3] test test_kernel.test.skip... SKIP
1 passed; 1 skipped; 1 failed.
error: the following test command failed with exit code 1:
zig-out/bin/zig-spirv-executor --platform Intel /home/robin/programming/projects/zig-spirv-test-runner/zig-cache/o/0010a1f2cc26bbadec8ba2fcdcbde1f8/test.spv
```
(dont worry, the fail and skip were intentional)

### CODE REVIEW

File: src/Compilation.zig [hunk 1]
```diff
@@ -1734,7 +1733,9 @@ pub fn create(gpa: Allocator, options: InitOptions) !*Compilation {
                 try comp.work_queue.writeItem(.{ .libssp = {} });
             }
 
-            if (!comp.bin_file.options.link_libc and capable_of_building_zig_libc) {
+            if (!comp.bin_file.options.link_libc and capable_of_building_zig_libc and
```
Comment (kubkon) [inline src/Compilation.zig:1736]: Could this be the culprit of the CI failures?

File: src/target.zig [hunk 2]
```diff
@@ -323,23 +323,26 @@ pub fn selfHostedBackendIsAsRobustAsLlvm(target: std.Target) bool {
 }
 
 pub fn supportsStackProbing(target: std.Target) bool {
-    return target.os.tag != .windows and target.os.tag != .uefi and
-        (target.cpu.arch == .x86 or target.cpu.arch == .x86_64);
+    return switch (target.cpu.arch) {
+        .x86, .x86_64 => target.os.tag != .windows and target.os.tag != .uefi,
```
Comment (kubkon) [inline src/target.zig:327]: I don't think this is correct. Prior to this change we would always return `false` if the os was `.windows` or `.uefi` regardless of the architecture.

File: src/target.zig [hunk 3]
```diff
@@ -323,23 +323,26 @@ pub fn selfHostedBackendIsAsRobustAsLlvm(target: std.Target) bool {
...
+        .spirv32, .spirv64 => false,
+        else => true,
```
Comment (jacobly0) [inline src/target.zig:329]: To preserve the original logic:
```suggestion
        else => false,
```

File: src/Compilation.zig (hunk 1 above)
Comment (Snektron) [reply src/Compilation.zig]: Ah yes, i misinterpreted the difference between libc and zig_libc.

Comment (Snektron) [reply src/Compilation.zig]: Yea that was it

**Pull request merged**